index.d.ts

```diff
@@ -0,0 +1,70 @@
+import { ComponentClass, Component, StatelessComponent } from 'react';
+import { Store, Dispatch, ActionCreator } from 'redux';
+
+interface ComponentDecorator<TOriginalProps, TOwnProps> {
+  (component: StatelessComponent<TOriginalProps>|ComponentClass<TOriginalProps>): ComponentClass<TOwnProps>;
+}
+
+/**
+ * Following 5 functions cover all possible ways connect could be invoked
+ *
+ * - State: Redux state interface (the same one used by Store<S>)
+ * - TOwnProps: Props passed to the wrapping component
+ * - TStateProps: Result of MapStateToProps
+ * - TDispatchProps: Result of MapDispatchToProps
+ */
+export function connect<State, TOwnProps>(): ComponentDecorator<{ dispatch: Dispatch<State> }, TOwnProps>;
+
+export function connect<State, TOwnProps, TStateProps>(
+  mapStateToProps: FuncOrSelf<MapStateToProps<State, TOwnProps, TStateProps>>,
+): ComponentDecorator<TStateProps & { dispatch: Dispatch<State> }, TOwnProps>;
+
+export function connect<State, TOwnProps, TStateProps, TDispatchProps>(
+  mapStateToProps: FuncOrSelf<MapStateToProps<State, TOwnProps, TStateProps>>,
+  mapDispatchToProps: FuncOrSelf<MapDispatchToPropsFunction<State, TOwnProps, TDispatchProps> | MapDispatchToPropsObject & TDispatchProps>
+): ComponentDecorator<TStateProps & TDispatchProps, TOwnProps>;
+
+export function connect<State, TOwnProps, TDispatchProps>(
+  mapStateToProps: null,
+  mapDispatchToProps: FuncOrSelf<MapDispatchToPropsFunction<State, TOwnProps, TDispatchProps> | MapDispatchToPropsObject & TDispatchProps>
+): ComponentDecorator<TDispatchProps, TOwnProps>;
+
+export function connect<State, TOwnProps, TStateProps, TDispatchProps, TMergeProps>(
+  mapStateToProps: FuncOrSelf<MapStateToProps<State, TOwnProps, TStateProps>>,
+  mapDispatchToProps: FuncOrSelf<MapDispatchToPropsFunction<State, TOwnProps, TDispatchProps> | MapDispatchToPropsObject & TDispatchProps>,
+  mergeProps: MergeProps<TOwnProps, TStateProps, TDispatchProps, TMergeProps>,
+  options?: Options
+): ComponentDecorator<TStateProps & TDispatchProps, TOwnProps>;
+
+interface MapDispatchToPropsObject {
+  [name: string]: ActionCreator<any>;
+}
+
+interface MapStateToProps<State, TOwnProps, TStateProps> {
+  (state: State, ownProps: TOwnProps): TStateProps;
+}
+
+interface MapDispatchToPropsFunction<State, TOwnProps, TDispatchProps> {
+  (dispatch: Dispatch<State>, ownProps: TOwnProps): TDispatchProps;
+}
+
+interface MergeProps<TOwnProps, TStateProps, TDispatchProps, TMergeProps> {
+  (ownProps: TOwnProps, stateProps: TStateProps, dispatchProps: TDispatchProps): TMergeProps;
```

Comment:
The order of arguments is incorrect, the right one would be: `stateProps, dispatchProps, ownProps`